doc/tools/coqrst/notations/parsing.py

```diff
@@ -12,6 +12,14 @@
 
 from antlr4 import CommonTokenStream, InputStream
 from antlr4.error.ErrorListener import ErrorListener
+from antlr4.Recognizer import Recognizer
+
+# Only print ANTLR version warning once
+check_version = Recognizer.checkVersion
+def checkVersion_once(*args, **kwargs):
+    check_version(*args, **kwargs)
+    Recognizer.checkVersion = lambda *args, **kwargs: None
+Recognizer.checkVersion = checkVersion_once
```

Comment:
It would be a bit simpler to define a boolean variable "checkedVersion" and use an if statement inside check_version_once rather than updating `Recognized.checkVersion` twice.  The latter is a little harder to follow when you read the code.  Just a thought.

Reply:
That's non-trivial because where do you store the boolean variable? In the `Recognizer` class? Anyway, I tried to implement your suggestion and didn't manage to. IMHO Clément's code is fine as is.

Reply:
This won't work?  But not worth obcessing over.

```suggestion
checked_version = false
check_version = Recognizer.checkVersion
def checkVersion_once(*args, **kwargs):
    if not checked_version:
        # Using "Recognizer.checkVersion" would cause endless recursion
        check_version(*args, **kwargs)
        checked_version = true
Recognizer.checkVersion = checkVersion_once
```

Reply:
Indeed, that's almost character for character the code I tested.
This gives the error (once you fix the casing for `true` and `false`):
```
UnboundLocalError: local variable 'checked_version' referenced before assignment
```

But thinking about this again, I've now figured what was missing to make it work (a `global checked_version` declaration).

Reply:
FWIW I originally wrote this code, then replaced it with the new version because I found that easier to read :) Also, you don't need the boolean variable:

```
checkVersion = Recognizer.checkVersion
def checkVersion_once(*args, **kwargs):
    global checkVersion
    if checkVersion:
        checkVersion(*args, **kwargs)
        checkVersion = None
Recognizer.checkVersion = checkVersion_once
```

And the shed should be red, of course :)